### ISSUE TITLE
libharfbuzz.so.0 Error

### ISSUE THREAD

**Issue by ghost** [spolu/breach_core#73]
```
./breach

__AUTO_UPDATE_BUNDLE__/exo_browser/exo_browser: error while loading shared libraries: libharfbuzz.so.0: cannot open shared object file: No such file or directory
```

Using Ubuntu 12.04. Maybe this is a problem with my computer?

**Comment by spolu** [comment]
Never had this issue and actually never heard of that library. Are you running chrome/chromium on this machine?

**Comment by amit5148** [comment]
Getting same error.
Chrome is running on same machine.

**Comment by freeride-pawel** [comment]
Ubuntu 12.04 - same error. Chromium works.
Installed this lib from the source http://www.freedesktop.org/wiki/Software/HarfBuzz/
Then a new error shows up:
/home/user/Desktop/breach-v0.3.20-alpha.5-linux-x64/**AUTO_UPDATE_BUNDLE**/exo_browser/exo_browser: /lib/x86_64-linux-gnu/libudev.so.1: no version information available (required by /home/user/Desktop/breach-v0.3.20-alpha.5-linux-x64/**AUTO_UPDATE_BUNDLE**/exo_browser/exo_browser)
/home/user/Desktop/breach-v0.3.20-alpha.5-linux-x64/**AUTO_UPDATE_BUNDLE**/exo_browser/exo_browser: symbol lookup error: /home/user/Desktop/breach-v0.3.20-alpha.5-linux-x64/**AUTO_UPDATE_BUNDLE**/exo_browser/exo_browser: undefined symbol: g_type_class_adjust_private_offset

**Comment by hhirsch** [comment]
error while loading shared libraries: libudev.so.1: cannot open shared object file: No such file or directory

After some installing some libraries from source. I will try to fix this later when I have more time.

**Comment by Rylius** [comment]
Linux Mint 14 (should be about the same as Ubuntu 12.04 library wise) and running Chrome.

```
breach-v0.3.20-alpha.5-linux-x64/__AUTO_UPDATE_BUNDLE__/exo_browser/exo_browser: error while loading shared libraries: libharfbuzz.so.0: cannot open shared object file: No such file or directory
```

Compiling the library from source interestingly enough didn't do anything (and it's not available in the repository), I had to manually copy the `libharfbuzz.so.0` to `__AUTO_UPDATE_BUNDLE__/exo_browser/lib`. (maybe HarfBuzz isn't installing into the right path or something, haven't looked into that)

Afterwards it complains about missing `libudev.so.1`, which is not available in the repository either. Source is a part of systemd, and my setup seems to be too old to build that. `libudev.so.0` is installed however.
Copying `/lib/x86_64-linux-gnu/libudev.so.0` to `__AUTO_UPDATE_BUNDLE__/exo_browser/lib/libudev.so.1` leads to the symbol lookup error `undefined symbol: g_type_class_adjust_private_offset`.

**Comment by thecotne** [comment]
+1

**Comment by jiripospisil** [comment]
> Compiling the library from source interestingly enough didn't do anything (and it's not available in the repository), I had to manually copy the libharfbuzz.so.0 to **AUTO_UPDATE_BUNDLE**/exo_browser/lib. (maybe HarfBuzz isn't installing into the right path or something, haven't looked into that)

@Rylius You might need to run `sudo ldconfig` afterwards so that the linker can pick up on new libraries that are available.

**Comment by Rylius** [comment]
@mekishizufu Ah cheers, completely forgot about that one. Fixes `libharfbuzz.so.0` not being found after compiling from source.

**Comment by ghost** [comment]
Yep! That fixes that error, but it seems that another one has come up for me.

```
__AUTO_UPDATE_BUNDLE__/exo_browser/exo_browser: error while loading shared libraries: libudev.so.1: cannot open shared object file: No such file or directory
```

Update to Ubuntu 12.10 does not fix it. `Copying /lib/x86_64-linux-gnu/libudev.so.0` to `__AUTO_UPDATE_BUNDLE__/exo_browser/lib/libudev.so.1` gives `symbol lookup error undefined symbol: g_type_class_adjust_private_offset`.

**Comment by didasy** [comment]
+1 on Ubuntu 12.10

**Comment by arunn** [comment]
+1 on Ubuntu 12.04

**Comment by francisbrito** [comment]
+1 on Ubuntu 12.04.

Running Google Chrome.

**Comment by spolu** [comment]
Hi guys. This is really annoying. Ideally I'd be curious to see if that would work if one of you would build breach / ExoBrowser from source on your machine. Anyone willing to give it a try?

**Comment by ghost** [comment]
@spolu, Done. Still the same error.

**Comment by spolu** [comment]
So you get the same error running the content shell in chromium?

**Comment by ghost** [comment]
Yeah. I think I've built it correctly.

**Comment by spolu** [comment]
So content shell runs but but not the ExoBrowser?

**Comment by ghost** [comment]
ExoBrowser doe not run.

**Comment by spolu** [comment]
Can you provide the output of both the `content_shell` and the `exo_browser` executables ? Thanks!

**Comment by jiripospisil** [comment]
Hey guys, so I actually managed to get it running on Ubuntu 12.04 yesterday. The way I did that is definitely not something you should ever do. It involves a lot of version renaming and it just gotta explode on you at some point. It's just to get it running and look around. Anyway, the quick way to get it running is to use the libraries I compiled and uploaded.

```
cd /tmp
mkdir breach_testing && cd breach_testing
wget https://raw.githubusercontent.com/breach/releases/master/v0.3.20-alpha.5/breach-v0.3.20-alpha.5-linux-x64.tar.gz
tar xfz breach-v0.3.20-alpha.5-linux-x64.tar.gz && cd breach-v0.3.20-alpha.5-linux-x64
wget https://dl.dropboxusercontent.com/u/4695780/libs.tar.gz
tar xfz libs.tar.gz -C __AUTO_UPDATE_BUNDLE__/exo_browser/lib
cp /opt/google/chrome/chrome-sandbox .
sudo chown root:root chrome-sandbox
sudo chmod 4755 chrome-sandbox
CHROME_DEVEL_SANDBOX="./chrome-sandbox" ./breach
```

If you want to compile the libraries yourself, then compile&install `libharfbuzz` and head `glib2` (to some temporary directory, you will probably need to install some dependencies as well). Copy the compiled `libharfbuzz` library into `lib` in the breach subdirectory. Do the same for `glib2`, you only need `libglib-2.*.so.*` and `libgobject-2.*.so.*`, that's where the missing symbols come from. Finally copy `libudev` from `/usr/lib/x86_64-linux-gnu`. Find the `exo_browser` executable, look at the output of `ldd exo_browser` and rename the libraries in `lib` to match. The recipe for `chrome-sandbox` is the same as above.

I suspect the steps may not work for everybody, I use this machine for native development all the time and may have some libraries installed you don't. And again, nobody should ever do it this way, you're on your own if things break.

**Comment by spolu** [comment]
Awesome! We definitely have to make Breach work on Ubuntu 12.04 out of the box. 
Or at least identify the package that need to be installed there.

**Comment by ghost** [comment]
Ubuntu 12.10, updated exobrowser to latest version.

[4765:4765:0717/123441:3482836470:FATAL:browser_main_loop.cc(172)] Running without the SUID sandbox! See https://code.google.com/p/chromium/wiki/LinuxSUIDSandboxDevelopment for more information on developing with the sandbox on.
./breach: line 6:  4765 Aborted                 (core dumped) $SRC_DIR/**AUTO_UPDATE_BUNDLE**/exo_browser/exo_browser --raw $SRC_DIR/**AUTO_UPDATE_BUNDLE**/breach_core

**Comment by spolu** [comment]
Yep see #60

**Comment by ghost** [comment]
Awesome :smile: Got it running on Ubuntu 12.10. I was experiencing all the above errors before.

![screenshot from 2014-07-17 13 31 50](https://cloud.githubusercontent.com/assets/6678142/3619139/836a9a06-0df1-11e4-912f-cfb9ee5182f2.png)

**Comment by spolu** [comment]
:+1:

**Comment by ghost** [comment]
@spolu @mekishizufu  Thanks for the steps, it worked for me too. I am running a Elementary distro(Luna) based on Ubuntu 12.04.

**Comment by spolu** [comment]
This should be solved by the release of v0.3.22-alpha.6
Linux x64: http://bit.ly/1p8Jkmc

Make sure to check README in tarball or https://github.com/breach/breach_core/wiki/Running-Breach-on-Linux
